### PR TITLE
Provide new assigns to fn in assign_new

### PR DESCRIPTION
**Pull request by joshchernoff**
Would like input. 🙏
The idea is simply to provide access to the new assigns in the fn of assign_new 

Example. 
```
assigns = assigns
        |> assign_new(:foo, fn  -> "foo")
        |> assign_new(:bar, fn %{foo: foo} -> foo <> "bar")
```

### CODE REVIEW

**Comment by joshchernoff** [comment]
This adds the concern that when using assigns with assign_new that you will need to ignore the assigns in the fn if not used.

File: lib/phoenix_live_view.ex
Comment (leandrocp) [inline lib/phoenix_live_view.ex:764]: Hi @joshchernoff unfortunately this line is introducing a breaking change, so I believe what you want is to actually add a new function clause like you did `assign_new(...) when is_function(fun, 1)` and keep the current one unchanged, in other words have support for both. Also update docs to explain and add an example of this new feature :)

And considering that test/phoenix_live_view_test.exs have the new test "has access to new assigns", then the other changes on that file and also the change on test/phoenix_component_test.exs may not be needed.

File: test/phoenix_live_view_test.exs
Comment (leandrocp) [inline test/phoenix_live_view_test.exs:315]:
```suggestion
      assigns =
        @assigns_changes
        |> assign_new(:another, fn _ -> "changed" end)
        |> assign_new(:and_another, fn %{another: another} -> another end)
```

**Comment by simonmcconnell** [comment]
Can you make it so we can pass either a 0 or 1 arity function? As it stands, this will break everyone's code.

**Review by leandrocp — COMMENTED** [review]
_Just my 2 cents trying to help_. Makes sense @josevalim ?

**Comment by leandrocp** [comment]
Close #1996

File: test/phoenix_live_view_test.exs
Comment (joshchernoff) [reply test/phoenix_live_view_test.exs]: Agreed I should made this additive and kept their prior 0 arity.

Comment (leandrocp) [reply test/phoenix_live_view_test.exs]: There are other `assign_new` functinos with `is_function(fun, 0)`guards that _may_ need their new pair `assign_new(...) when is_function(fun, 1)`. You can double check by playing with existing tests and passing fun/1 to them.

Comment (joshchernoff) [reply test/phoenix_live_view_test.exs]: Thought about only providing the fun/1 when using assigns only and not a socket at first?

I ask because I'm unsure about the right approch to. 
```
 %{private: %{assign_new: {assigns, keys}}} ->
        # It is important to store the keys even if they are not in assigns
        # because maybe the controller doesn't have it but the view does.
        socket = put_in(socket.private.assign_new, {assigns, [key | keys]})
        Phoenix.LiveView.Utils.force_assign(socket, key, Map.get_lazy(assigns, key, fun))
```
In the context of using a socket.

Comment (leandrocp) [reply test/phoenix_live_view_test.exs]: That's a API design decision so it's not really up to me to say for sure, but as a LV user I'd expect it to just work when passing either a `socket` or `assigns` considering the first arg of `assign_new` is `socket_or_assigns`.

I guess you're asking about the line `Phoenix.LiveView.Utils.force_assign(socket, key, Map.get_lazy(assigns, key, fun))`, you can just write it as:

```elixir
Phoenix.LiveView.Utils.force_assign(socket, key, Map.get_lazy(assigns, key, fn -> fun.(socket.assigns) end))
```

But feel free to wait for the feedback of @josevalim or someone from the core team :)

Comment (joshchernoff) [reply test/phoenix_live_view_test.exs]: I agree that it would be better to also support sockets, I was just unsure about how. I will try this.

Comment (leandrocp) [reply test/phoenix_live_view_test.exs]: Just an update, it probably should be `Phoenix.LiveView.Utils.force_assign(socket, key, Map.get_lazy(assigns, key, fn -> fun.(assigns) end))` instead.

Comment (joshchernoff) [reply test/phoenix_live_view_test.exs]: Yeah as I suspected the `Map.get_lazy` does not like the new arg. 

```
pry(1)>  Map.get_lazy(assigns, key, fun.(assigns))
** (FunctionClauseError) no function clause matching in Map.get_lazy/3

    The following arguments were given to Map.get_lazy/3:

        # 1
        %{existing: "existing-parent"}

        # 2
        :existing3

        # 3
        "existing-parent"

    Attempted function clauses (showing 1 out of 1):

        def get_lazy(map, key, fun) when is_function(fun, 0)

    (elixir 1.13.2) lib/map.ex:491: Map.get_lazy/3
    (stdlib 3.16.1) erl_eval.erl:685: :erl_eval.do_apply/6
    (iex 1.13.2) lib/iex/evaluator.ex:310: IEx.Evaluator.handle_eval/3
    (iex 1.13.2) lib/iex/evaluator.ex:285: IEx.Evaluator.do_eval/3
    (iex 1.13.2) lib/iex/evaluator.ex:274: IEx.Evaluator.eval/3
    (iex 1.13.2) lib/iex/evaluator.ex:169: IEx.Evaluator.loop/1
    (iex 1.13.2) lib/iex/evaluator.ex:32: IEx.Evaluator.init/4
    (phoenix_live_view 0.17.9) lib/phoenix_live_view.ex:764: Phoenix.LiveView.assign_new/3
    test/phoenix_live_view_test.exs:316: Phoenix.LiveViewUnitTest."test assign_new with socket has access to assigns"/1
    (ex_unit 1.13.2) lib/ex_unit/runner.ex:500: ExUnit.Runner.exec_test/1
    (stdlib 3.16.1) timer.erl:166: :timer.tc/1
    (ex_unit 1.13.2) lib/ex_unit/runner.ex:451: anonymous fn/4 in ExUnit.Runner.spawn_test_monitor/4
 ```

Comment (leandrocp) [reply test/phoenix_live_view_test.exs]: Yep `Map.get_lazy/3` expect a fun 👍🏻 

And a test like this should pass:

```elixir
    test "with 1 arity function" do
      socket =
        @socket
        |> assign_new(:another, fn _ -> "changed" end)
        |> assign_new(:and_another, fn %{another: another} -> another end)

      assert socket.assigns == %{
               another: "changed",
               and_another: "changed",
               live_action: nil,
               flash: %{},
               __changed__: %{another: true, and_another: true}
             }
    end
```

Comment (joshchernoff) [reply test/phoenix_live_view_test.exs]: Before I continue I think its worth discussing the expecting behavior for the socket.  
in the case of https://github.com/phoenixframework/phoenix_live_view/blob/46132e5a8b1ab81f612e5b59eecab53b30da5a06/test/phoenix_live_view_test.exs#L277-L294

You end up here. https://github.com/phoenixframework/phoenix_live_view/blob/46132e5a8b1ab81f612e5b59eecab53b30da5a06/lib/phoenix_live_view.ex#L753-L757

In which case the assigns on the socket have yet to be updated and the new assigns are stored in the private's assign_new.

IE. If I inspect the socket I will see `existing` had not changed while the private's new assigns for existing has. 
```
  assigns: %{
    __changed__: %{existing: true, existing2: true},
    existing: "existing-parent",
    existing2: "existing2",
    flash: %{},
    live_action: nil
  },
  
  vs in sockets private.
  
  %{
  __changed__: %{},
  assign_new: {%{existing: "existing-parent"}, [:existing]},
  connect_info: %{},
  connect_params: %{},
  root_view: Phoenix.LiveViewTest.ParamCounterLive
}
```

Whats the best approach to this?

Comment (leandrocp) [reply test/phoenix_live_view_test.exs]: Just to make sure I follow your question.

For the case of the following calls:

```elixir
       |> assign_new(:existing, fn assigns -> "new-existing" end) 
       |> assign_new(:existing2, fn assigns -> "new-existing2" end) 
```

None of those two functions will be called because `:existing` and `:existing2` keys are already declared, so it's fine.

I guess your question is about the last one where the fn would be called passing something as argument:

```elixir
|> assign_new(:notexisting, fn assigns -> "new-notexisting" end)
```

So your concern is what `assigns` should return?



> I guess something like this would be expected?

The expected assigns on that example, IMO should be:

```elixir
existing: "existing-parent", # parent already defined, do not overwrite
existing2: "existing2", # defined on assign/2
existing_parent: "existing-parent" # current value of `existing`
existing3: "existing2" # current value of `existing2`
````

_And no `existing_parent_updated`_

The problem is currently you only see the `existing` assign which is defined on `@socket.private[:assign_new]` so it won't work. _If_ that's the correct expected result then it needs a bit of extra logic to fetch all assigns before passing to `fn -> fun.(assigns) end`. Makes sense?

Sorry I can't say for sure if that's the expected behavior and if it should be implemented, it's just what I'd expect If calling that function. So it's better to wait for some feedback, ok?

Comment (joshchernoff) [reply test/phoenix_live_view_test.exs]: > IMO the expected assigns on that example should be:
> 
> ```elixir
> existing: "existing-parent", # parent already defined, do not overwrite
> existing2: "existing2", # defined on assign/2
> existing_parent: "existing-parent" # current value of `existing`
> existing3: "existing2" # current value of `existing2`
> ```

Ah I think I miss read the test assuming ` |> assign_new(:existing, fn -> "new-existing" end)` overriding the prior `existing-parent` and I would have access to that.

**Comment by 03juan** [comment]
This is a great feature, well done on the effort! I think it would be helpful if it also worked for `update/3`.

**Comment by josevalim** [comment]
The changes are in the correct direction IMO! We just need to fix warnings, tests, and we are good to go!

@03juan if you want to send a similar PR for update, please do!

**Review by leandrocp — COMMENTED** [review]
Hi @joshchernoff just some tips following https://github.com/phoenixframework/phoenix_live_view/pull/1997#issuecomment-1111971779. Feel free to ask :)

File: test/phoenix_live_view_test.exs
Comment (joshchernoff) [reply test/phoenix_live_view_test.exs:330]: Q: Are you intently ignoring in the case of `fn _ -> "new-existing" end)` to cause the fun/1 ?
I just assumed it would call the fun/0 and that you could drop the ignored.

File: lib/phoenix_live_view.ex
Comment (joshchernoff) [reply lib/phoenix_live_view.ex]: I think I was following along, pushed update.

Comment (joshchernoff) [reply lib/phoenix_live_view.ex]: Bummer, looks like I've failed on 1.9.4. Will need to fix that.

Comment (joshchernoff) [reply lib/phoenix_live_view.ex]: Ok looks like a false positive, but still maybe worth looking at in another ticket.

File: lib/phoenix_live_view.ex
Comment (leandrocp) [inline lib/phoenix_live_view.ex:747]:
```suggestion
  Note that `fun` also provides access to the previously assigned values:

      assigns =
          assigns
          |> assign_new(:foo, fn -> "foo")
```

Just nitpick to fix "asscess"  and format the code.

**Review by leandrocp — APPROVED** [review]
LGTM!

Just one last suggestion to improve docs 👍🏻

File: lib/phoenix_live_view.ex
Comment (leandrocp) [reply lib/phoenix_live_view.ex]: All tests are green, it should be good.

File: lib/phoenix_live_view.ex
Comment (joshchernoff) [reply lib/phoenix_live_view.ex]: Thanks for that, my spelling is horrible at times

Comment (leandrocp) [reply lib/phoenix_live_view.ex]: No problem 👍🏻

File: lib/phoenix_live_view.ex
Comment (simonmcconnell) [inline lib/phoenix_live_view.ex:749]:
```suggestion
          |> assign_new(:foo, fn -> "foo" end)
          |> assign_new(:bar, fn %{foo: foo} -> foo <> "bar" end)
```

Comment (joshchernoff) [reply lib/phoenix_live_view.ex]: good catch.

File: lib/phoenix_live_view.ex
Comment (leandrocp) [inline lib/phoenix_live_view.ex:766]: Another solution to clean up `socket.assigns` is to call https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.Helpers.html#assigns_to_attributes/2, see https://github.com/phoenixframework/phoenix_live_view/pull/1999#discussion_r862013992

I forgot about that function but seems like a good approach. WDYT?

Comment (joshchernoff) [reply lib/phoenix_live_view.ex:766]: I'm not so sure thats better. At the moment I'm seeing that I have to provide a list of ignores which means I have to figure out which to ignore first.  I gave it a spin and also learned it returns a Keyword list which also means I have to manage that when merging the two.  Did you have a specific thought in mind?  its not inherently coming to me as I'm trying it.

Comment (leandrocp) [reply lib/phoenix_live_view.ex:766]: Never mind, I was thinking about reusing that function but it won't work.

File: lib/phoenix_live_view.ex
Comment (josevalim) [inline lib/phoenix_live_view.ex:774]:
```suggestion

        Phoenix.LiveView.Utils.force_assign(
          socket,
          key,
          case assigns do
            %{^key => value} -> value
            %{} -> fun.(socket.assigns)
          end
        )
```

Comment (josevalim) [reply lib/phoenix_live_view.ex:774]: We don't want to merge those assigns because the assigns in `assigns_new` are assigns that have not been moved into the socket yet.

File: lib/phoenix_live_view.ex
Comment (josevalim) [inline lib/phoenix_live_view.ex:773]:
```suggestion
        Phoenix.LiveView.Utils.force_assign(
          socket,
          key,
          case assigns do
            %{^key => value} -> value
            %{} -> fun.(socket.assigns)
          end
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

File: lib/phoenix_live_view.ex
Comment (joshchernoff) [reply lib/phoenix_live_view.ex:772]: thanks 🙏

**Comment by joshchernoff** [comment]
I just wanted to say this was my first contribution to live view and I have really been enjoying this lib. 
So thank you all for helping me and providing to this pr 🎉

**Comment by leandrocp** [comment]
> I just wanted to say this was my first contribution to live view and I have really been enjoying this lib.
> So thank you all for helping me and providing to this pr 🎉

Thanks for your contribution!

**Pull request merged**
